modules/core/src/main/java/org/apache/ignite/cache/query/index/Index.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.cache.query.index;
+
+import java.util.UUID;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.internal.processors.cache.persistence.CacheDataRow;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Basic interface for Ignite indexes.
+ */
+public interface Index {
+    /**
+     * Unique ID.
+     */
+    public UUID id();
+
+    /**
+     * Index name.
+     */
+    public String name();
+
+    /**
+     * Checks whether index handles specified cache row.
+     *
+     * @param row Cache row.
+     * @return Whether index handles specified cache row
+     */
+    public boolean belongsToIndex(CacheDataRow row) throws IgniteCheckedException;
+
+    /**
+     * Callback that runs when the underlying cache is updated.
+     *
+     * @param oldRow Cache row that was replaced with newRow.
+     * @param newRow Cache row that was stored.
+     * @param prevRowAvailable Whether oldRow available.
+     */
+    public void onUpdate(@Nullable CacheDataRow oldRow, @Nullable CacheDataRow newRow, boolean prevRowAvailable)
+        throws IgniteCheckedException;
+
+    /**
+     * Put cache row to index.
+     *
+     * @return {@code True} if replaced existing row.
+     */
+    public boolean putx(CacheDataRow row) throws IgniteCheckedException;
```

Comment:
The previous implementation of H2TreeIndex used the `putx` method only. It did have the `put` method, but in all places it's not used returned value, just checks it for existing, so it's similar to `putx` logic. So it looks like it's just not required. I can add it to make API more wide, but as it won't be a part of public for a while API, do we need it?

Reply:
well, I always interpret the `-x` suffix as `extended` (for example, with or without additional checks), and to have an extended version of something, you have to have the general one first. But it's just my intuition. And with these in mind I'm OK to leave this as is. Furthermore, since it will be moved to the internal space, we will be able to rename it late if needed.